accepted/0042-isolated-mode.md

```diff
@@ -0,0 +1,347 @@
+
+## Motivation
+
+The introduction of workspaces in npm v7 brought the ability to split large code-bases into smaller pieces of code with declared dependencies between them.
+
+Several build tools took advantage of these declared dependencies to split the builds into multiple steps (one per workspace). These build tools take advantage of the declared dependencies between workspaces to only rebuild the workspaces that have not changed since the last build. With the current way npm works, any dependency change to any workspace may affect any other workspace. This means that the build systems need to rebuild the entire project every time a single workspace modifies a dependency. This is inefficient and does not scale.
+
+## Naming
+
+This mode is called `isolated-mode` because it makes the boundaries between packages and workspaces stronger, making them more isolated. 
+
+
+To make the discussions more efficient, we could also name the current default installation mode of npm. We will call this mode `hoisted-mode` as packages are shared across workspaces by hoisting them to the root of the project. This term is already known by developers familiar with yarn or pnpm.
+
+## More on the problem
+
+Packages and workspaces declare their dependencies to npm by using special fields in their `package.json`. The Node.js runtime is unaware of these special fields, instead it uses a [module resolution algorithm](https://nodejs.org/api/modules.html) to determine the meaning of a module importing another module. npm bridges this gap by converting the declared dependency graph into a folder structure which makes sense to Node.js. This operation is call `reification`.
+
+When converting the dependency graph to a folder structure, `hoisted-mode` loses information. Multiple different dependency graphs can be converted to an identical folder structure. Since the folder structure is all that Node.js uses to resolve modules, Node.js misses some information about dependencies. This loss of information is most frequently visible through the fact that workspaces are able to successfully import the dependencies declared by other workspaces.
+
+### Forgetting to declare dependencies
+
+When workspaces can successfully use code of a package without having a dependency on it, people forget to declare their dependencies. This leads to situations where updating the dependencies of one workspace breaks a seemingly unrelated workspace.
+
+It is worth noting that static code analysis tools can help significantly reduce the frequency of these mistakes. While static code analysis tools are probably sufficient for small projects, they are not for large projects. "isolated-mode" not only helps get the dependencies right but also _proves_ that the dependencies are correct. This proof is essential to build trustable build systems which use caching, scoped installations.
```

Comment:
Why are they not for large projects? I’ve used many extremely large projects with eslint-plugin-import and it works just fine.

Reply:
Static analysis tools are not enough guarantee to be able to have a build system reliably do caching  (not re-building everything when only a small part of the repo changes). This is because a dependency of one workspace can break another unrelated workspace. Static code analysis does not remove this risk.

Reply:
In my experience, static code analysis tools are sufficient to force all dependencies to be explicitly declared, in large as well as small projects, which seems like it obviates the problem you're describing.

Reply:
one example that we hit is a workspace adding a dependency to `@types/jasmine` and that breaks all the other workspaces which use `@types/jest`. Because `@types/jasmine` will be hoisted and included in the compilation of all the workspaces, making them fail. This cannot be caught by code analysis tools because the dependencies are actually correctly declared.

Reply:
ok, that makes sense - maybe instead of "they are not", a more nuanced sentence could be used, since it's really not the size of the project that causes the issue, it's "there's more than one child workspace".

It also might be unique to packages with TS types, or to these specific packages which have colliding type definitions - do you have any examples of this happening with JS?

Reply:
another example:
- `workspace 1` depends on `foo`
- `foo` requires `yargs` but does not declare it as a dependency
- `workspace 2` depends on `yargs`
I all works fine, foo can require `yargs` because it is hoisted.

- now `workspace 2` get rid of `yargs`
- `workspace 1` is broken because `foo` cannot import `yargs` anymore

In this situation all the dependencies are correctly declared in my repo, so it would not be caught by code analysis.

Reply:
> maybe instead of "they are not", a more nuanced sentence could be used, since it's really not the size of the project that causes the issue, it's "there's more than one child workspace".

It is not limited to repositories using workspaces, in the example above, if you replace `workspace 2` by an external package called `bar` and you make `workspace 1` take a dependency on `bar` you get the same issue.

It is actually not limited to the size of the repo, but the likelihood of this being an issue increases with the complexity of the dependency graph, to a point a which it becomes unmanageable passed a certain scale (which we have passed in internal repos in Microsoft).

Reply:
Sure - but the likelihood isn’t zero in a small repo either, so the size of the repo doesn’t flip a binary, it instead increases a probability - and it’d be great for the prose here to reflect that.